docs/adrs/003-Retry_endpoint_design.md

```diff
@@ -0,0 +1,162 @@
+* The retry will only continue if the status of the job to be retried is in [Status.terminated.value, Status.error.value]
+* If the job id points to a job that has already been retried, it will attempt to retry that job's `retry_parent` instead.
+* If the job id has never been retried, it becomes the `retry_parent` 
+* EE2 looks up the method versions and parameters, and then submits the job to be retried, incrementing the `retry_count` of the job being retried, and the newly launched job gains a field called `retry_parent` that contains the job id of the job from the original request.
+* The job is submitted and upon successful submission, the child job adds the field `retry_parent` and notifies the `parent_job_id` that a new `child_job` has been added by appending itself to the `parent_job.child_jobs[]` field
+* There is no way to specify ResourceRequirements with a retry at the moment, even if the job was previously submitted by an administrator and had specfified ResourceRequirements. The retry will only use resource requirements from the catalog / ee2 config.
+
+
+### Batch Behavior
+* If a job has the attribute of `batch_job=True` the retry will fail, since there is no method to re-run. This is a bug, as it doesn't fail gracefully.
+* If a job has the attribute of `batch_job=True`, but is actually a child job, the parent will be notified of this new retried job
+* Multiple in-flight retries are allowed.
+* Adds `child_job_id` to `parent_job_id.child_job_ids[]`
+
+## Retry_job behavior
+* Blocking and single submit to HTCondor. It should be fine as it returns relatively quickly
+  
+## Retry_jobs behavior
+* Submitting multiple jobs for retry serially calls the same code path
+used for running a single job and blocks until all jobs have been
+submitted to the condor queue. This can cause issues if the
+network drops, and makes the narrative not aware of the state of
+the retry. Submitting 100 jobs currently takes 9 seconds, and that
+is a lot of time for things to go wrong. 
+* (Follow up: Hopefully the making the narrative aware of the state of the retry will be mitigated by the narrative backend. It just blocks on the call anyway, with the default service timeout, which I think is something wacky like half an hour. As long as the user doesn't kill the kernel at that time, all should be well. Of course, if it were me, and it looked frozen for more than a couple minutes, I'd probably restart. )
+* Multiple in-flight retries are allowed.
+
+### Desired Behavior
+
+#### General
+* Prevent multiple in-flight retries to prevent the user from wasting their own resources (and the queues resources)
+* Non blocking job submission for submitting multiple jobs, possibly via using `run_job_batch` (requires refactor of run_job_batch)
+* One single submission to HTCondor instead of multiple job submissions
```

Comment:
We're doing the two above for normal jobs, but it's not clear to me that we decided to do them for retry jobs - did we do that? Personally, I'm a lot less worried about it for retries than normal jobs...

Reply:
Deciding to do them for run_job mostly does it for retry_job, as run_job is used by retry_job at some point.
* Non blocking submit is https://kbase-jira.atlassian.net/browse/DATAUP-535
* Submitting multiple retry jobs needs to be done with `run_multiple` but isn't hooked up yet, and also affects error reporting https://github.com/kbase/execution_engine2/pull/406/files
* All jobs are doing single submissions to htcondor right now

Reply:
We're I'm coming from is that we do need to speed up regular job submission as described above. I don't think it needs to be done for retry job (although it'd be nice), and I don't think we decided that we needed to do it, so we should document that in the ADR somewhere, which right now I don't see.

If it's easy and gets done, great, but let's not make that a requirement.

Reply:
Yea it doesn't matter so much for a retry_job, but it does matter for retrying large numbers of jobs. I moved the ticket into the non mvp section

Reply:
:+1: